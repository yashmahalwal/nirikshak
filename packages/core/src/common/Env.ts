/* istanbul ignore file */
/* eslint-disable prefer-const */
export let MAX_PLURAL_ENTRIES = 10;
export let MIN_PLURAL_ENTRIES = 0;
export let RANDOMNESS_ITERATIONS = 10;
