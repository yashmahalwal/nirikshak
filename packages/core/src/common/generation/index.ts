export { generateFaker } from "./fakerGen";
export { generateCustom } from "./customGen";
