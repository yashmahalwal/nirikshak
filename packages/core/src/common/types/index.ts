export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
