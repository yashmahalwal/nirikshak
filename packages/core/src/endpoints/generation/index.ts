export { generateURL } from "./urlStringGen";
export { getFromResource } from "./resourceStringGen";
export * from "./helpers";
export * from "./input";
