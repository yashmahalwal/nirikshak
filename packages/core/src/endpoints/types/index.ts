export {
    MethodType,
    HeaderMap,
    Cases,
    BodyType,
    BodyInstance,
} from "./helpers";
export { isURLString } from "./urlString";
export { isResourceString } from "./resourceString";
export { isDescription, Description } from "./description";
export { HeaderAndQuery, Inputs } from "./input";
export { HeaderAndStatus, Outputs } from "./output";
