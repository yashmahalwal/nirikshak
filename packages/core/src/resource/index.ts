export * from "./generation";
export * from "./types";
