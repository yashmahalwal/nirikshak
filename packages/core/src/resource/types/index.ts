import {
    isBaseType,
    isWithModifiersBaseType,
    isWithModifiersResource,
    isOneOfEntries,
    BaseType,
    WithModifiers,
    OneOfEntries,
} from "./helper";

// type guard: ResourceBase
export function isResourceBase(input: any): input is ResourceBase {
    if (!input || typeof input != "object" || Array.isArray(input))
        return false;
    return Object.keys(input).every(
        (key) =>
            isBaseType(input[key]) ||
            isWithModifiersBaseType(input[key]) ||
            isWithModifiersResource(input[key]) ||
            isOneOfEntries(input[key]) ||
            isResourceBase(input[key])
    );
}

// A basic key value pair
export interface ResourceBase {
    [key: string]:
        | BaseType
        | WithModifiers<BaseType>
        | WithModifiers<ResourceBase>
        | OneOfEntries
        // Check for resource base at last as it matches any key
        | ResourceBase;
}

// A resource schema is a resource base along with an identifier field
export interface Resource extends ResourceBase {
    id: Exclude<BaseType, Array<any> | null | boolean>;
}

// type guard: isResource
export function isResource(input: any): input is Resource {
    if (!input || typeof input !== "object") return false;

    return (
        !Array.isArray(input["id"]) &&
        isBaseType(input.id) &&
        isResourceBase(input)
    );
}
