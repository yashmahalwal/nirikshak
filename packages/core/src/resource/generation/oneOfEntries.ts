import faker from "faker";
import { BaseType, WithModifiers, OneOfEntries } from "../types/helper";
import { ResourceBase } from "../types/resource";

// Return type with discriminated union
export type SelectedType<T extends BaseType | ResourceBase> = {
    type: T extends BaseType ? "base type" : "resource base";
    selection: T extends BaseType
        ? BaseType | WithModifiers<BaseType>
        : ResourceBase | WithModifiers<ResourceBase>;
};

// Function simply selects one of the types / fields
// and returns
export function generateOneOfEntries(
    input: OneOfEntries
): SelectedType<BaseType> | SelectedType<ResourceBase> {
    // Use faker.random.number instead of Math.random
    // So that randomness seed works

    // From the given selections, select an entry

    // If only types exists
    // Or we randomly choose type from choice b/w types and fields
    if (!("fields" in input))
        // Only type in the input
        return {
            type: "base type",
            selection:
                input.types[
                    faker.random.number({
                        min: 0,
                        max: input.types.length - 1,
                        precision: 1,
                    })
                ],
        };

    // If only fields exists
    if (!("types" in input))
        return {
            type: "resource base",
            selection:
                input.fields[
                    faker.random.number({
                        min: 0,
                        max: input.fields.length - 1,
                        precision: 1,
                    })
                ],
        };

    // Select either type or field
    return faker.random.boolean()
        ? {
              type: "resource base",
              selection:
                  input.fields[
                      faker.random.number({
                          min: 0,
                          max: input.fields.length - 1,
                          precision: 1,
                      })
                  ],
          }
        : {
              type: "base type",
              selection:
                  input.types[
                      faker.random.number({
                          min: 0,
                          max: input.types.length - 1,
                          precision: 1,
                      })
                  ],
          };
}
