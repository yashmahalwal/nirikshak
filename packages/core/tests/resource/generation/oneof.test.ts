import { OneOfEntries } from "../../../src/resource/types";
import faker from "faker";
import { RANDOMNESS_ITERATIONS } from "../../../src/resource/Env";
import { generateOneOfEntries } from "../../../src/resource/generation/oneOfEntries";

const entries: OneOfEntries[] = [
    {
        types: [
            1,
            2,
            3,
            ["stringify", 19, 30, 19.542],
            "faker:random.number",
            [[[[["custom.username"]]]]],
        ],
    },
    {
        fields: [
            { branch: "CSE" },
            {
                age: 12,
                name: 14,
                address: {
                    types: [
                        12.4353,
                        "faker:random.number",
                        { type: true, nullable: true },
                        [12, 13, 14, null, false, "myString"],
                    ],
                },
            },
        ],
    },
    {
        fields: [
            { name: "Name", surName: "Surname" },
            {
                age: {
                    type: {
                        function: "faker:random.number",
                        args: [{ min: 1, max: 100, precision: 0.1 }],
                    },
                    nullable: true,
                },
            },
        ],
        types: [12, 14, true, true, false, ["SIMP"]],
    },
];

describe("Oneof tests", () => {
    beforeAll(() => faker.seed(123));

    for (let i = 0; i < RANDOMNESS_ITERATIONS; i++)
        entries.forEach((entry, index) =>
            test(`Entry ${index}, iteration: ${i}`, () => {
                const o = generateOneOfEntries(entry);
                // Validate with snapshot
                expect(o).toMatchSnapshot();
                // Check if the entry came from the input
                expect(
                    ("types" in entry &&
                        entry.types.includes(o.selection as any)) ||
                        ("fields" in entry &&
                            entry.fields.includes(o.selection as any))
                ).toBe(true);
            })
        );
});
