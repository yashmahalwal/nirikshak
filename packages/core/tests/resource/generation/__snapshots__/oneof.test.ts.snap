// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Oneof tests Entry 0, iteration: 0 1`] = `
Object {
  "selection": "faker:random.number",
  "type": "base type",
}
`;

exports[`Oneof tests Entry 0, iteration: 1 1`] = `
Object {
  "selection": 2,
  "type": "base type",
}
`;

exports[`Oneof tests Entry 0, iteration: 2 1`] = `
Object {
  "selection": "faker:random.number",
  "type": "base type",
}
`;

exports[`Oneof tests Entry 0, iteration: 3 1`] = `
Object {
  "selection": Array [
    Array [
      Array [
        Array [
          Array [
            "custom.username",
          ],
        ],
      ],
    ],
  ],
  "type": "base type",
}
`;

exports[`Oneof tests Entry 0, iteration: 4 1`] = `
Object {
  "selection": 3,
  "type": "base type",
}
`;

exports[`Oneof tests Entry 0, iteration: 5 1`] = `
Object {
  "selection": 3,
  "type": "base type",
}
`;

exports[`Oneof tests Entry 0, iteration: 6 1`] = `
Object {
  "selection": 3,
  "type": "base type",
}
`;

exports[`Oneof tests Entry 0, iteration: 7 1`] = `
Object {
  "selection": 3,
  "type": "base type",
}
`;

exports[`Oneof tests Entry 0, iteration: 8 1`] = `
Object {
  "selection": 2,
  "type": "base type",
}
`;

exports[`Oneof tests Entry 0, iteration: 9 1`] = `
Object {
  "selection": Array [
    "stringify",
    19,
    30,
    19.542,
  ],
  "type": "base type",
}
`;

exports[`Oneof tests Entry 1, iteration: 0 1`] = `
Object {
  "selection": Object {
    "address": Object {
      "types": Array [
        12.4353,
        "faker:random.number",
        Object {
          "nullable": true,
          "type": true,
        },
        Array [
          12,
          13,
          14,
          null,
          false,
          "myString",
        ],
      ],
    },
    "age": 12,
    "name": 14,
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 1, iteration: 1 1`] = `
Object {
  "selection": Object {
    "address": Object {
      "types": Array [
        12.4353,
        "faker:random.number",
        Object {
          "nullable": true,
          "type": true,
        },
        Array [
          12,
          13,
          14,
          null,
          false,
          "myString",
        ],
      ],
    },
    "age": 12,
    "name": 14,
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 1, iteration: 2 1`] = `
Object {
  "selection": Object {
    "branch": "CSE",
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 1, iteration: 3 1`] = `
Object {
  "selection": Object {
    "branch": "CSE",
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 1, iteration: 4 1`] = `
Object {
  "selection": Object {
    "branch": "CSE",
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 1, iteration: 5 1`] = `
Object {
  "selection": Object {
    "address": Object {
      "types": Array [
        12.4353,
        "faker:random.number",
        Object {
          "nullable": true,
          "type": true,
        },
        Array [
          12,
          13,
          14,
          null,
          false,
          "myString",
        ],
      ],
    },
    "age": 12,
    "name": 14,
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 1, iteration: 6 1`] = `
Object {
  "selection": Object {
    "branch": "CSE",
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 1, iteration: 7 1`] = `
Object {
  "selection": Object {
    "address": Object {
      "types": Array [
        12.4353,
        "faker:random.number",
        Object {
          "nullable": true,
          "type": true,
        },
        Array [
          12,
          13,
          14,
          null,
          false,
          "myString",
        ],
      ],
    },
    "age": 12,
    "name": 14,
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 1, iteration: 8 1`] = `
Object {
  "selection": Object {
    "branch": "CSE",
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 1, iteration: 9 1`] = `
Object {
  "selection": Object {
    "address": Object {
      "types": Array [
        12.4353,
        "faker:random.number",
        Object {
          "nullable": true,
          "type": true,
        },
        Array [
          12,
          13,
          14,
          null,
          false,
          "myString",
        ],
      ],
    },
    "age": 12,
    "name": 14,
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 2, iteration: 0 1`] = `
Object {
  "selection": true,
  "type": "base type",
}
`;

exports[`Oneof tests Entry 2, iteration: 1 1`] = `
Object {
  "selection": Object {
    "age": Object {
      "nullable": true,
      "type": Object {
        "args": Array [
          Object {
            "max": 100,
            "min": 1,
            "precision": 0.1,
          },
        ],
        "function": "faker:random.number",
      },
    },
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 2, iteration: 2 1`] = `
Object {
  "selection": false,
  "type": "base type",
}
`;

exports[`Oneof tests Entry 2, iteration: 3 1`] = `
Object {
  "selection": Object {
    "age": Object {
      "nullable": true,
      "type": Object {
        "args": Array [
          Object {
            "max": 100,
            "min": 1,
            "precision": 0.1,
          },
        ],
        "function": "faker:random.number",
      },
    },
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 2, iteration: 4 1`] = `
Object {
  "selection": true,
  "type": "base type",
}
`;

exports[`Oneof tests Entry 2, iteration: 5 1`] = `
Object {
  "selection": Object {
    "name": "Name",
    "surName": "Surname",
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 2, iteration: 6 1`] = `
Object {
  "selection": false,
  "type": "base type",
}
`;

exports[`Oneof tests Entry 2, iteration: 7 1`] = `
Object {
  "selection": Object {
    "age": Object {
      "nullable": true,
      "type": Object {
        "args": Array [
          Object {
            "max": 100,
            "min": 1,
            "precision": 0.1,
          },
        ],
        "function": "faker:random.number",
      },
    },
  },
  "type": "resource base",
}
`;

exports[`Oneof tests Entry 2, iteration: 8 1`] = `
Object {
  "selection": 12,
  "type": "base type",
}
`;

exports[`Oneof tests Entry 2, iteration: 9 1`] = `
Object {
  "selection": Object {
    "name": "Name",
    "surName": "Surname",
  },
  "type": "resource base",
}
`;
